File: leetcode2976/solution.cpp
/*
Solution 1:

Floyd-Warshall Algorithm

Time: O(n + C) | Space: O(n ^ 2)

- n: length of edges
- C: the maximum value of C is (26 ^ 3 + 25 ^ 2)
*/

class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        vector<vector<int>> costs (26, vector<int> (26, -1));

        // the cost is 0 if the letter doesn't change
        for (int i = 0; i < 26; ++i) {
            costs[i][i] = 0;
        }

        // initialize costs
        for (int i = 0; i < cost.size(); ++i) {
            int from = original[i] - 'a';
            int to = changed[i] - 'a';

            if (costs[from][to] == -1) {
                costs[from][to] = cost[i];
            } else {
                costs[from][to] = min(costs[from][to], cost[i]);
            }

        }

        // floyd-warshall
        for (int k = 0; k < 26; ++k) {
            for (int i = 0; i < 26; ++i) {
                for (int j = 0; j < 26; ++j) {
                    if (costs[i][k] == -1 || costs[k][j] == -1) {
                        continue;
                    }

                    if (costs[i][j] == -1) {
                        costs[i][j] = costs[i][k] + costs[k][j];
                    } else {
                        costs[i][j] = min(costs[i][j], costs[i][k] + costs[k][j]);
                    }
                }
            }
        }

        long long res = 0;

        for (int i = 0; i < source.size(); ++i) {
            int from = source[i] - 'a';
            int to = target[i] - 'a';

            if (costs[from][to] == -1) {
                return -1;
            }

            res += costs[from][to];
        }
        
        return res;
    }
};

File: leetcode2976/solution.rs
/*
Solution 1:

Floyd-Warshall Algorithm

Time: O(n + C) | Space: O(n ^ 2)

- n: length of edges
- C: the maximum value of C is (26 ^ 3 + 25 ^ 2)
*/

impl Solution {
    pub fn minimum_cost(source: String, target: String, original: Vec<char>, changed: Vec<char>, cost: Vec<i32>) -> i64 {
        let mut costs: Vec<Vec<i64>> = vec![vec![-1; 26]; 26];

        // the cost is 0 if the letter doesn't change
        for i in 0..26 {
            costs[i][i] = 0;
        }

        // initialize costs
        for i in 0..cost.len() {
            let from: usize = ((original[i] as u8) - ('a' as u8)) as usize;
            let to: usize = ((changed[i] as u8) - ('a' as u8)) as usize;

            costs[from][to] = if costs[from][to] == -1 {
                cost[i] as i64
            } else {
                costs[from][to].min(cost[i] as i64)
            };
        }

        // floyd-warshall
        for k in 0..26 {
            for i in 0..26 {
                for j in 0..26 {
                    if costs[i][k] == -1 || costs[k][j] == -1 {
                        continue;
                    }
                    
                    costs[i][j] = if costs[i][j] == -1 {
                        costs[i][k] + costs[k][j]
                    } else {
                        costs[i][j].min(costs[i][k] + costs[k][j])
                    };
                }
            }
        }

        let mut res: i64 = 0;

        for (from, to) in source.bytes().zip(target.bytes()) {
            let from: usize = (from - ('a' as u8)) as usize;
            let to: usize = (to - ('a' as u8)) as usize;

            if costs[from][to] == -1 {
                return -1;
            }

            res += costs[from][to];
        }

        res
    }
}